contrib/shs/shs.1
.\"	BSDI	$Id: shs.1,v 1.1.1.1 1994/01/13 22:42:33 polk Exp $
.\" @(#)shs.1	6.1 1/10/94 22:52:34
.\"
.\" This man page was written by:
.\"
.\"	 Landon Curt Noll  (chongo@toad.com)	chongo <was here> /\../\
.\"
.\" This man page has been placed in the public domain.  Please do not 
.\" copyright this man page.
.\"
.\" LANDON CURT NOLL DISCLAIMS ALL WARRANTIES WITH  REGARD  TO
.\" THIS  SOFTWARE,  INCLUDING  ALL IMPLIED WARRANTIES OF MER-
.\" CHANTABILITY AND FITNESS.  IN NO EVENT SHALL  LANDON  CURT
.\" NOLL  BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL
.\" DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM  LOSS  OF
.\" USE,  DATA  OR  PROFITS, WHETHER IN AN ACTION OF CONTRACT,
.\" NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR  IN
.\" CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\" See shsdrvr.c for version and modification history.
.\"
.TH SHS 1  LOCAL
.SH NAME
shs \- NIST proposed Secure Hash Standard
.SH SYNOPSIS
.BR shs \ [
.BR \-cdhiqtx \ ][
.BR \-p prefix\ ][
.BR \-P prfile\ ][
.BR \-s str\ ]
file\ ...
.SH DESCRIPTION
The
.B shs
utility implements the Secure Hash Algorithm (\fISHA\fP).
It produces 160-bit Secure Hash Digests of files, strings or data 
read on stdin.
If no
.B file
is given, and
no string is to be digested (\fB\-s\fP)
then stdin will be digested.
By default, a digests are printed as 40 hex characters without a leading 
.IR 0x .
.PP
On January 31, 1992, NIST published in the Federal Register a proposed
Secure Hash Standard (\fISHS\fP) which is based on 
.IR SHA .
The 
.I SHS
is designed to work with the proposed Digital Signature Standard (\fIDSS\fP).
The Secure Hash Algorithm (\fISHA\fP) 
specified in this standard is
necessary to ensure the security of the Digital Signature Standard. 
When a message of length < 2^64 bits is input, the 
.I SHA 
produces a 160-bit representation of the message called the message digest. 
The message digest is used during generation of a signature for the message.  
The
.I SHA 
is designed to have the following properties: it is
computationally infeasible to recover a message corresponding to a
given message digest, or to find two different messages which
produce the same message digest.
.PP
If a 
.B str 
(string) argument is given, then the digest for 
.BR str ,
followed by a space, followed by
.B str 
enclosed double quotes is written to stdout.
Any
.B file
arguments are ignored.
.PP
If one or more
.B file
if given, a separate digest if produced for each file.
By default, 
.B file 
digests are followed by a space and the filename.
.PP
If no 
.B str
or
.B file
arguments are given, then a digest of stdin is written to stdout.
.PP
.TP
.B \-c
Print C style hex digests with a leading \fI0x\fP.
.TP
.B \-d
Compute dual digests.
.sp
Data is divided into two streams and digested separately.
Each successive octet is assigned an index starting with 0.
The even stream consists of octets with even indices.
The odd stream consists of octets with odd indices.
The even stream digest, followed by a space, followed by
the odd stream digest is written to stdout.
.TP
.B \-h
Print a help and usage message.
.TP
.B \-i
Compute inode digests of files.
.sp
Prepend the filename and various inode information to the file data 
being digested.
The inode information prepended includes information
such as the device (\fIst_dev\fP), 
inode number (\fIst_ino\fP), 
mode (\fIst_mode\fP), 
link count (\fIst_nlink\fP), 
uid (\fIst_uid\fP), 
gid (\fIst_gid\fP), 
size (\fIst_size\fP),
modification time (\fIst_mtime\fP) and
change time (\fIst_ctime\fP).
The prepended data is padded with zeros to make it a multiple of 64
bytes long.
Both a
.BR stat (2)
and a
.BR lstat (2)
information are both used.
.sp
The \fB\-i\fP flag allows one to include various inode information in
the digest.  
This option is useful in detecting file tampering.
For example, the following will produce different digests:
.sp
.in +0.5i
.nf
shs \-i /tmp/chongo
cp /tmp/chongo /tmp/was_here
\^... misc funny business ...
cp /tmp/was_here /tmp/chongo
shs \-i /tmp/chongo
.fi
.in -0.5i
.sp
Note that a \fB\-i\fP digest is portable to other systems.
This is because inode information will likely change as the
contents of a file are copied from machine to machine.
.sp
To mark a \fB\-i\fP digest as a special value, \fI.0\fP is 
prepended onto the digest output.
.sp
One may only compute inode digests of files.
Use of
.B \-i
disables reading from stdin.
The
.B \-i
flag is not compatible with
.BR \-s .
.TP
.BI \-p prefix
.TP
.BI \-P prfile
Insert a prefix into the data to be digested.
.sp
By using \fB\-p\fP\fIprefix\fP, one may prepend any set of data with a
string.  
The digest produced is equivalent to digest that is
produced with the string pretended to the data.  
Thus the following two commands produce the same digest:
.sp
.in +0.5i
.nf
shs \-p curds \-s whey
shs \-s curdswhey
.fi
.in -0.5i
.sp
By use of the \fB\-P\fP\fIprfile\fP interface, one may prepend using
up to the 32k of a file.  
This interface allows one to prepend using binary data.  
The following produces the same digest:
.sp
.in +0.5i
.nf
shs \-P /usr/bin/awk /bin/ls
dd if=/usr/bin/awk of=/tmp/foo bs=32k count=1
cat /tmp/foo /bin/ls > /tmp/ls
shs /tmp/ls
.fi
.in -0.5i
.sp
Knowledge of the original prepend data is only kept in
the digest.  
Thus one may use the prepend string as a ``salt'' making it intractable 
for someone else to reproduce the digest of a file without knowledge 
of the prepend data.
.sp
For sightly faster performance, use prepend data that is
a multiple of 64 bytes long.  
The easy way to do this is to use \fB-P\fP on a file that is at least 32k
bytes in length.
.sp
This feature works in conjunction all modes of operation except
the \fB\-t\fP and \fB\-x\fP modes.
.TP
.B \-q
Output only digests.
The filename or string will not be written to stdout.
.TP
.BR \-s str
Digest \fIstr\fP as if it were a string.
.sp
The trailing NUL byte is not digested.
No files are digested.
The \fIstr\fP is written to stdout enclosed in double quotes.
.TP
.BR \-t
Time the user cpu seconds needed to digest several megabytes of data.
The number of megabytes,
followed by the digest produced,
followed by the number of user cpu seconds
followed by the number of characters per user second is written to stdout.
.sp
On some systems with certain virtual memory characteristics,
you may need to run the performance test several times, ignoring the
first result.
The default amount of data tested is 16 megabytes.
On some systems, this amount may differ.
One may change the number of megabytes processed by changing the value of
TEST_MEG in the file shsdrvr.c and recompiling.
.TP
.BR \-v
Print the version.
.TP
.BR \-x
Perform an extended standard SHS test suite.
.sp
The test suite will first look in the current directory for
.IR file1 .
If it is not found, it will look in
.IR /usr/contrib/lib/shs .
It is assumes that 
.I file2 
is in the same location as 
.sp
The standard test suite been extended.  
The initial string now says:
.sp
.in +0.5i
.nf
shs test suite results
.fi
.in -0.5i
.sp
instead of:
.sp
.in +0.5i
.nf
SHA test suite results
.fi
.in -0.5i
.sp
to reflect the utility name instead of the algorithm implemented.
This also helps distinguish this version from much older 
versions which did not have the 
.B -v 
flag.
Also the original test file
.I foo
which contained the string  "\fIabc\fP" (with no newline) was renamed 
.IR file1 .
.SH "SEE ALSO"
.BR md5 (1),
.BR stat (2),
.BR lstat (2)
.SH FILES
.nf
\^./file1			default test file location
\^./file2			default test file location
/usr/contrib/lib/shs/file1	alternate test file location
/usr/contrib/lib/shs/file2	alternate test file location
.fi
.SH AUTHOR
.nf
Much of this code was written, re-written or modified by:

    Landon Curt Noll	(chongo@toad.com)  /\\../\\

This code is based on code by Peter C. Gutmann.  Much thanks goes 
to Peter C. Gutman (pgut1@cs.aukuni.ac.nz) , Shawn A. Clifford
(sac@eng.ufl.edu), Pat Myrto (pat@rwing.uucp), Colin Plumb
(colin@nyx10.cs.du.edu), Rich Schroeppel (rcs@cs.arizona.edu)
and others who wrote and/or worked on the original code.
.fi
.SH NOTICE
LANDON CURT NOLL DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. 
IN NO EVENT SHALL LANDON CURT NOLL BE LIABLE FOR ANY SPECIAL, INDIRECT OR
CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF
USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
.SH BUGS
The command:
.sp
.in +0.5i
.nf
shs -s '/\\"O/\\'
.fi
.in -0.5i
.sp
writes to stdout, the following line:
.sp
.in +0.5i
.nf
013d51217d98b6f2cfa4e7ea13bad8f9bc506f23 "/\\"O/\\"
.fi
.in -0.5i
.sp
The string written in double quotes by
.B \-s
may not be a valid string according to C or shell syntax.


contrib/shs/shs.c
/*	BSDI	$Id: shs.c,v 1.1.1.1 1994/01/13 22:42:33 polk Exp $	*/
/* @(#)shs.c	6.3 1/11/94 01:06:50 */
/*
 * shs - NIST proposed Secure Hash Standard
 *
 * Written 2 September 1992, Peter C. Gutmann.
 *
 * This file was Modified/Re-written by:
 *
 *	 Landon Curt Noll  (chongo@toad.com)	chongo <was here> /\../\
 *
 * This code has been placed in the public domain.  Please do not 
 * copyright this code.
 *
 * LANDON CURT NOLL DISCLAIMS ALL WARRANTIES WITH  REGARD  TO
 * THIS  SOFTWARE,  INCLUDING  ALL IMPLIED WARRANTIES OF MER-
 * CHANTABILITY AND FITNESS.  IN NO EVENT SHALL  LANDON  CURT
 * NOLL  BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL
 * DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM  LOSS  OF
 * USE,  DATA  OR  PROFITS, WHETHER IN AN ACTION OF CONTRACT,
 * NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR  IN
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * See shsdrvr.c for version and modification history.
 */

#include <string.h>
#include "shs.h"
#ifdef __bsdi__
#include <machine/endian.h>
#else
#include "align.h"
#include "endian.h"
#endif

char *shs_what="@(#)";	/* #(@) if checked in */

/* 
 * The SHS f()-functions.  The f1 and f3 functions can be optimized 
 * to save one boolean operation each - thanks to Rich Schroeppel,
 * rcs@cs.arizona.edu for discovering this.
 *
 * f1: ((x&y) | (~x&z)) == (z ^ (x&(y^z)))
 * f3: ((x&y) | (x&z) | (y&z)) == ((x&y) | (z&(x|y)))
 */
#define f1(x,y,z)       (z ^ (x&(y^z)))		/* Rounds  0-19 */
#define f2(x,y,z)       (x^y^z)			/* Rounds 20-39 */
#define f3(x,y,z)       ((x&y) | (z&(x|y)))	/* Rounds 40-59 */
#define f4(x,y,z)       (x^y^z)			/* Rounds 60-79 */
 
/* The SHS Mysterious Constants */
#define K1      0x5A827999L                                                                     /* Rounds  0-19 */
#define K2      0x6ED9EBA1L                                                                     /* Rounds 20-39 */
#define K3      0x8F1BBCDCL                                                                     /* Rounds 40-59 */
#define K4      0xCA62C1D6L                                                                     /* Rounds 60-79 */
 
/* SHS initial values */
#define h0init  0x67452301L
#define h1init  0xEFCDAB89L
#define h2init  0x98BADCFEL
#define h3init  0x10325476L
#define h4init  0xC3D2E1F0L
 
/* 32-bit rotate left - kludged with shifts */
#define ROT(X,n)  (((X)<<(n)) | ((X)>>(32-(n))))
 
/* 
 * The initial expanding function.  The hash function is defined over an
 * 80-word expanded input array W, where the first 16 are copies of the input
 * data, and the remaining 64 are defined by
 *
 *      W[ i ] = W[ i - 16 ] ^ W[ i - 14 ] ^ W[ i - 8 ] ^ W[ i - 3 ]
 *
 * This implementation generates these values on the fly in a circular
 * buffer - thanks to Colin Plumb, colin@nyx10.cs.du.edu for this
 * optimization.
 */
#define exor(W,i) (W[(i)&15] ^= W[(i)-14&15] ^ W[(i)-8&15] ^ W[(i)-3&15])

/* 
 * The prototype SHS sub-round.  The fundamental sub-round is:
 *
 *      a' = e + ROT(a,5) + f(b,c,d) + k + data;
 *      b' = a;
 *      c' = ROT(b,30);
 *      d' = c;
 *      e' = d;
 *
 * but this is implemented by unrolling the loop 5 times and renaming the
 * variables ( e, a, b, c, d ) = ( a', b', c', d', e' ) each iteration.
 * This code is then replicated 20 times for each of the 4 functions, using
 * the next 20 values from the W[] array each time.
 */
#define subRound(a, b, c, d, e, f, k, data) \
    (e += ROT(a,5) + f(b,c,d) + k + data, b = ROT(b,30))

/* forward declarations */
static void shsTransform P((ULONG*, ULONG*));


/*
 * Initialize the SHS values
 */
void
shsInit(shsInfo)
    SHS_INFO *shsInfo;
{
    /* Set the h-vars to their initial values */
    shsInfo->digest[0] = h0init;
    shsInfo->digest[1] = h1init;
    shsInfo->digest[2] = h2init;
    shsInfo->digest[3] = h3init;
    shsInfo->digest[4] = h4init;

    /* Initialise bit count */
    shsInfo->countLo = shsInfo->countHi = shsInfo->datalen = 0L;
}


/*
 * Perform the SHS transformation.  Note that this code, like MD5, seems to
 * break some optimizing compilers - it may be necessary to split it into
 * sections, eg based on the four subrounds
 */
static void
shsTransform(digest, W)
    ULONG *digest;
    ULONG *W;
{
    ULONG A, B, C, D, E;	/* Local vars */
 
    /* Set up first buffer and local data buffer */
    A = digest[0];
    B = digest[1];
    C = digest[2];
    D = digest[3];
    E = digest[4];
 
    /* Heavy mangling, in 4 sub-rounds of 20 interations each. */
    subRound(A, B, C, D, E, f1, K1, W[ 0]);
    subRound(E, A, B, C, D, f1, K1, W[ 1]);
    subRound(D, E, A, B, C, f1, K1, W[ 2]);
    subRound(C, D, E, A, B, f1, K1, W[ 3]);
    subRound(B, C, D, E, A, f1, K1, W[ 4]);
    subRound(A, B, C, D, E, f1, K1, W[ 5]);
    subRound(E, A, B, C, D, f1, K1, W[ 6]);
    subRound(D, E, A, B, C, f1, K1, W[ 7]);
    subRound(C, D, E, A, B, f1, K1, W[ 8]);
    subRound(B, C, D, E, A, f1, K1, W[ 9]);
    subRound(A, B, C, D, E, f1, K1, W[10]);
    subRound(E, A, B, C, D, f1, K1, W[11]);
    subRound(D, E, A, B, C, f1, K1, W[12]);
    subRound(C, D, E, A, B, f1, K1, W[13]);
    subRound(B, C, D, E, A, f1, K1, W[14]);
    subRound(A, B, C, D, E, f1, K1, W[15]);
    subRound(E, A, B, C, D, f1, K1, exor(W,16));
    subRound(D, E, A, B, C, f1, K1, exor(W,17));
    subRound(C, D, E, A, B, f1, K1, exor(W,18));
    subRound(B, C, D, E, A, f1, K1, exor(W,19));
 
    subRound(A, B, C, D, E, f2, K2, exor(W,20));
    subRound(E, A, B, C, D, f2, K2, exor(W,21));
    subRound(D, E, A, B, C, f2, K2, exor(W,22));
    subRound(C, D, E, A, B, f2, K2, exor(W,23));
    subRound(B, C, D, E, A, f2, K2, exor(W,24));
    subRound(A, B, C, D, E, f2, K2, exor(W,25));
    subRound(E, A, B, C, D, f2, K2, exor(W,26));
    subRound(D, E, A, B, C, f2, K2, exor(W,27));
    subRound(C, D, E, A, B, f2, K2, exor(W,28));
    subRound(B, C, D, E, A, f2, K2, exor(W,29));
    subRound(A, B, C, D, E, f2, K2, exor(W,30));
    subRound(E, A, B, C, D, f2, K2, exor(W,31));
    subRound(D, E, A, B, C, f2, K2, exor(W,32));
    subRound(C, D, E, A, B, f2, K2, exor(W,33));
    subRound(B, C, D, E, A, f2, K2, exor(W,34));
    subRound(A, B, C, D, E, f2, K2, exor(W,35));
    subRound(E, A, B, C, D, f2, K2, exor(W,36));
    subRound(D, E, A, B, C, f2, K2, exor(W,37));
    subRound(C, D, E, A, B, f2, K2, exor(W,38));
    subRound(B, C, D, E, A, f2, K2, exor(W,39));

    subRound(A, B, C, D, E, f3, K3, exor(W,40));
    subRound(E, A, B, C, D, f3, K3, exor(W,41));
    subRound(D, E, A, B, C, f3, K3, exor(W,42));
    subRound(C, D, E, A, B, f3, K3, exor(W,43));
    subRound(B, C, D, E, A, f3, K3, exor(W,44));
    subRound(A, B, C, D, E, f3, K3, exor(W,45));
    subRound(E, A, B, C, D, f3, K3, exor(W,46));
    subRound(D, E, A, B, C, f3, K3, exor(W,47));
    subRound(C, D, E, A, B, f3, K3, exor(W,48));
    subRound(B, C, D, E, A, f3, K3, exor(W,49));
    subRound(A, B, C, D, E, f3, K3, exor(W,50));
    subRound(E, A, B, C, D, f3, K3, exor(W,51));
    subRound(D, E, A, B, C, f3, K3, exor(W,52));
    subRound(C, D, E, A, B, f3, K3, exor(W,53));
    subRound(B, C, D, E, A, f3, K3, exor(W,54));
    subRound(A, B, C, D, E, f3, K3, exor(W,55));
    subRound(E, A, B, C, D, f3, K3, exor(W,56));
    subRound(D, E, A, B, C, f3, K3, exor(W,57));
    subRound(C, D, E, A, B, f3, K3, exor(W,58));
    subRound(B, C, D, E, A, f3, K3, exor(W,59));
 
    subRound(A, B, C, D, E, f4, K4, exor(W,60));
    subRound(E, A, B, C, D, f4, K4, exor(W,61));
    subRound(D, E, A, B, C, f4, K4, exor(W,62));
    subRound(C, D, E, A, B, f4, K4, exor(W,63));
    subRound(B, C, D, E, A, f4, K4, exor(W,64));
    subRound(A, B, C, D, E, f4, K4, exor(W,65));
    subRound(E, A, B, C, D, f4, K4, exor(W,66));
    subRound(D, E, A, B, C, f4, K4, exor(W,67));
    subRound(C, D, E, A, B, f4, K4, exor(W,68));
    subRound(B, C, D, E, A, f4, K4, exor(W,69));
    subRound(A, B, C, D, E, f4, K4, exor(W,70));
    subRound(E, A, B, C, D, f4, K4, exor(W,71));
    subRound(D, E, A, B, C, f4, K4, exor(W,72));
    subRound(C, D, E, A, B, f4, K4, exor(W,73));
    subRound(B, C, D, E, A, f4, K4, exor(W,74));
    subRound(A, B, C, D, E, f4, K4, exor(W,75));
    subRound(E, A, B, C, D, f4, K4, exor(W,76));
    subRound(D, E, A, B, C, f4, K4, exor(W,77));
    subRound(C, D, E, A, B, f4, K4, exor(W,78));
    subRound(B, C, D, E, A, f4, K4, exor(W,79));
 
    /* Build message digest */
    digest[0] += A;
    digest[1] += B;
    digest[2] += C;
    digest[3] += D;
    digest[4] += E;
}


/*
 * Update SHS for a block of data.  This code does not assume that the
 * buffer size is a multiple of SHS_BLOCKSIZE bytes long.  This code
 * handles partial blocks between calls to shsUpdate().
 *
 * Changed to deal with partial buffers.  This in turn fixed the
 * time trial results.		Landon Curt Noll -- chongo@toad.com /\../\
 */
void
shsUpdate(shsInfo, buffer, count)
    SHS_INFO *shsInfo;
    BYTE *buffer;
    UINT count;
{
    ULONG in[SHS_BLOCKSIZE/sizeof(ULONG)];
    ULONG datalen = shsInfo->datalen;
    ULONG value;
    long tmp;
    int cnt;

    /*
     * Update bitcount
     */
    tmp = shsInfo->countLo;
    if ((shsInfo->countLo = tmp + ((ULONG)count<<3)) < tmp ) {
        shsInfo->countHi++;                             
    }
    shsInfo->countHi += count >> 29;

    /*
     * Catch the case of a non-empty data buffer
     */
    if (datalen > 0) {

	/* determine the size we need to copy */
	ULONG cpylen = SHS_BLOCKSIZE - datalen;

	/* case: new data will not fill the buffer */
	if (cpylen > count) {
	    memcpy((char *)(shsInfo->data)+datalen, buffer, count);
	    shsInfo->datalen = datalen+count;
	    return;

	/* case: buffer will be filled */
	} else {
	    memcpy((char *)(shsInfo->data)+datalen, buffer, cpylen);
#if BYTE_ORDER == LITTLE_ENDIAN 
	    /* byte swap data into big endian order */
	    for (cnt=0; cnt < SHS_BLOCKSIZE/sizeof(ULONG); ++cnt) {
		value = (shsInfo->data[cnt] << 16) |
			(shsInfo->data[cnt] >> 16);
		in[cnt] = ((value & 0xFF00FF00L) >> 8) |
			  ((value & 0x00FF00FFL) << 8);
	    }
	    shsTransform(shsInfo->digest, in);
#else
	    shsTransform(shsInfo->digest, shsInfo->data);
#endif
	    buffer += cpylen;
	    count -= cpylen;
	    shsInfo->datalen = 0;
	}
    }

    /*
     * Process data in SHS_BLOCKSIZE chunks
     */
    while (count >= SHS_BLOCKSIZE) {
#if defined(MUST_ALIGN)
	memcpy(in, buffer, SHS_BLOCKSIZE);
#if BYTE_ORDER == LITTLE_ENDIAN 
	/* byte swap data into big endian order */
	for (cnt=0; cnt < SHS_BLOCKSIZE/sizeof(ULONG); ++cnt) {
	    value = (in[cnt] << 16) | (in[cnt] >> 16);
	    in[cnt] = ((value & 0xFF00FF00L) >> 8) |
		      ((value & 0x00FF00FFL) << 8);
	}
#endif
	shsTransform(shsInfo->digest, in);
#else
#if BYTE_ORDER == BIG_ENDIAN
	shsTransform(shsInfo->digest, (ULONG *)buffer);
#else
	/* byte swap data into big endian order */
	for (cnt=0; cnt < SHS_BLOCKSIZE/sizeof(ULONG); ++cnt) {
	    value = (((ULONG *)buffer)[cnt] << 16) | 
	    	    (((ULONG *)buffer)[cnt] >> 16);
	    in[cnt] = ((value & 0xFF00FF00L) >> 8) |
		      ((value & 0x00FF00FFL) << 8);
	}
	shsTransform(shsInfo->digest, in);
#endif
#endif
	buffer += SHS_BLOCKSIZE;
	count -= SHS_BLOCKSIZE;
    }

    /*
     * Handle any remaining bytes of data.
     * This should only happen once on the final lot of data
     */
    if (count > 0) {
	memcpy(shsInfo->data, buffer, count);
    }
    shsInfo->datalen = count;
}


/*
 * Update SHS for a block of data with no partial block.  This function
 * assumes that count is a multiple of SHS_BLOCKSIZE and that no partial
 * blocks is left over from a previous call.
 */
void
shsfullUpdate(shsInfo, buffer, count)
    SHS_INFO *shsInfo;
    BYTE *buffer;
    UINT count;
{
#if defined(MUST_ALIGN) || BYTE_ORDER == LITTLE_ENDIAN
    ULONG in[SHS_BLOCKSIZE/sizeof(ULONG)];
#endif
    ULONG value;
    int cnt;

    /*
     * Update bitcount
     */
    if ((shsInfo->countLo + ((ULONG) count << 3)) < shsInfo->countLo) {
	 /* Carry from low to high bitCount */
	 shsInfo->countHi++;
    }
    shsInfo->countLo += ((ULONG) count << 3);
    shsInfo->countHi += ((ULONG) count >> 29);

    /*
     * Process data in SHS_BLOCKSIZE chunks
     */
    while (count >= SHS_BLOCKSIZE) {
#if defined(MUST_ALIGN)
	memcpy(in, buffer, SHS_BLOCKSIZE);
#if BYTE_ORDER == LITTLE_ENDIAN 
	/* byte swap data into big endian order */
	for (cnt=0; cnt < SHS_BLOCKSIZE/sizeof(ULONG); ++cnt) {
	    value = (in[cnt] << 16) | (in[cnt] >> 16);
	    in[cnt] = ((value & 0xFF00FF00L) >> 8) |
		      ((value & 0x00FF00FFL) << 8);
	}
#endif
	shsTransform(shsInfo->digest, in);
#else
#if BYTE_ORDER == BIG_ENDIAN
	shsTransform(shsInfo->digest, (ULONG *)buffer);
#else
	/* byte swap data into big endian order */
	for (cnt=0; cnt < SHS_BLOCKSIZE/sizeof(ULONG); ++cnt) {
	    value = (((ULONG *)buffer)[cnt] << 16) | 
	    	    (((ULONG *)buffer)[cnt] >> 16);
	    in[cnt] = ((value & 0xFF00FF00L) >> 8) |
		      ((value & 0x00FF00FFL) << 8);
	}
	shsTransform(shsInfo->digest, in);
#endif
#endif
	buffer += SHS_BLOCKSIZE;
	count -= SHS_BLOCKSIZE;
    }
}


void
shsFinal(shsInfo)
    SHS_INFO *shsInfo;
{
    int cnt;
    ULONG value;
    int count = shsInfo->datalen;
    ULONG lowBitcount = shsInfo->countLo;
    ULONG highBitcount = shsInfo->countHi;

    /*
     * Set the first char of padding to 0x80.
     * This is safe since there is always at least one byte free
     */
    ((BYTE *)shsInfo->data)[count++] = 0x80;

    /* Pad out to 56 mod 64 */
    if (count > 56) {
	/* Two lots of padding:  Pad the first block to 64 bytes */
	memset((BYTE *) shsInfo->data + count, 0, 64 - count);
#if BYTE_ORDER == LITTLE_ENDIAN 
	/* byte swap data into big endian order */
	for (cnt=0; cnt < SHS_BLOCKSIZE/sizeof(ULONG); ++cnt) {
	    value = (shsInfo->data[cnt] << 16) |
		    (shsInfo->data[cnt] >> 16);
	    shsInfo->data[cnt] = ((value & 0xFF00FF00L) >> 8) |
				 ((value & 0x00FF00FFL) << 8);
	}
#endif
	shsTransform(shsInfo->digest, shsInfo->data);

	/* Now fill the next block with 56 bytes */
	memset(shsInfo->data, 0, 56);
    } else
	/* Pad block to 56 bytes */
	memset((BYTE *) shsInfo->data + count, 0, 56 - count);
#if BYTE_ORDER == LITTLE_ENDIAN
	/* byte swap data into big endian order */
	for (cnt=0; cnt < SHS_BLOCKSIZE/sizeof(ULONG); ++cnt) {
	    value = (shsInfo->data[cnt] << 16) |
	    	    (shsInfo->data[cnt] >> 16);
	    shsInfo->data[cnt] = ((value & 0xFF00FF00L) >> 8) |
	    			 ((value & 0x00FF00FFL) << 8);
	}
#endif

    /*
     * Append length in bits and transform
     */
    shsInfo->data[14] = highBitcount;
    shsInfo->data[15] = lowBitcount;
    shsTransform(shsInfo->digest, shsInfo->data);
#if BYTE_ORDER != BIG_ENDIAN 
	for (cnt=0; cnt < SHS_DIGESTSIZE/sizeof(ULONG); ++cnt) {
	    value = (shsInfo->data[cnt] << 16) |
	    	    (shsInfo->data[cnt] >> 16);
	    shsInfo->data[cnt] = ((value & 0xFF00FF00L) >> 8) |
	    			 ((value & 0x00FF00FFL) << 8);
	}
#endif
}
